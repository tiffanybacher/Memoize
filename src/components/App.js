import React, { Component } from 'react';
import Nav from '../components/Nav';
import Cat from '../components/Cat';
import Instructions from '../components/Instructions'
import Flashcard from '../components/Flashcard';
import AnswerInput from '../components/AnswerInput';
import UserArea from '../components/UserArea';

class App extends Component {
  constructor() {
    super();
    this.state = {
      group: 1,
      allCards: '',
      cardSelection: this.getSavedCardSelection() || null,
      currentCardData: '',
      answerIsShown: false,
      userAnswer: '',
      savedMsgIsShown: false,
      savedCards: this.getSavedCards() || null
    }
  }

  componentDidMount() {
    fetch('https://fe-apps.herokuapp.com/api/v1/memoize/1901/tiffanybacher/objectquestions')
      .then(response => response.json())
      .then(data => this.setState({
        allCards: data.objectQuestions }))
      .then(data => this.startFreshCardSelection())
      .then(data => this.setState({
        currentCardData: this.state.cardSelection[0] }))
      .catch(err => console.log(err));
  }

  showAnswer = () => {
    this.setState({
      answerIsShown: true
    });
  }

  getUserAnswer = (answerInput) => {
    this.setState({
      userAnswer: answerInput
    });
  }

  getSavedCardSelection = () => {
    return JSON.parse(localStorage.getItem('savedSelection'));
  }

  startFreshCardSelection = () => {
    if (!this.state.cardSelection) {
      this.setState({
        cardSelection: this.state.allCards
      });
    }
  }

  updateCardSelection = () => {
    if (!this.state.cardSelection) {
      let length = this.state.allCards.length;

      this.setState({
        cardSelection: this.state.allCards.slice(1, length)
      });
    } else {
      let length = this.state.cardSelection.length;

      this.setState({
        cardSelection: this.state.cardSelection.slice(1, length)
      }, () => this.updateCardData());
    }
  }

  saveCurrentCardsToStorage = () => {
    if (this.state.cardSelection) {
      let savedSelection = JSON.parse(localStorage.getItem('savedSelection'));

      savedSelection = this.state.cardSelection;

      localStorage.setItem('savedSelection', JSON.stringify(savedSelection));
    }
  }

  updateCardData = () => {
    this.saveCurrentCardsToStorage();
    if (this.state.cardSelection) {
      this.setState({
        currentCardData: this.state.cardSelection[0]
      });
    } else {
      this.setState({
        currentCardData: this.state.allCards[0]
      });
    }
  }

  getSavedCards = () => {
    return JSON.parse(localStorage.getItem('savedCards'));
  }

  saveCardToStorage = () => {
    if (this.state.savedCards) {
      let allSavedCards = JSON.parse(localStorage.getItem('savedCards'));

      allSavedCards.push(this.state.allCards[0]);
      localStorage.setItem('savedCards', JSON.stringify(allSavedCards));
    } else {
      localStorage.setItem('savedCards', JSON.stringify([this.state.allCards[0]]));
    }
  }

  updateSavedCards = () => {
    if (!this.state.savedCards) {
      this.setState({
        savedCards: [this.state.currentCardData]
      });
    } else {
      this.setState({
        savedCards: this.state.savedCards.concat(this.state.currentCardData)
      })
    }
  }

  hideAnswer = () => {
    this.setState({
      answerIsShown: false
    });
  }

  showSavedMsg = () => {
    this.setState({
      savedMsgIsShown: true
    });
  }

  hideSavedMsg = () => {
    this.setState({
      savedMsgIsShown: false
    })
  }

  render() {
    // console.log('saved cards:', this.state.savedCards);
    // console.log('card selection:', this.state.cardSelection);
    // console.log('current card:', this.state.currentCardData);

    let userInput;

    if (this.state.answerIsShown) {
      userInput = 
        <UserArea 
          userAnswer={this.state.userAnswer}
          hideAnswer={this.hideAnswer}
          updateCardSelection={this.updateCardSelection}
          showSavedMsg={this.showSavedMsg}
          hideSavedMsg={this.hideSavedMsg}
          saveCardToStorage={this.saveCardToStorage}
          updateSavedCards={this.updateSavedCards}
        />
    } else {
      userInput = 
        <AnswerInput 
          showAnswer={this.showAnswer}
          getUserAnswer={this.getUserAnswer}
        />
    }

    return (
      <div className="App">
        <header className="App-header">
          <Nav />
        </header>
        <Cat />
        <section className="main-background">
          <div className="main-container">
            <Instructions
              answerIsShown={this.state.answerIsShown}
              savedMsgIsShown={this.state.savedMsgIsShown} />
            <Flashcard 
              question={this.state.currentCardData.question}
              answer={this.state.currentCardData.answer}
              answerIsShown={this.state.answerIsShown}
            />
            {userInput}
          </div>
        </section>
      </div>
    );
  }
}

export default App;
