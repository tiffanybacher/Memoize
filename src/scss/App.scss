html {
  margin: 0;
  padding: 0;
  min-width: 400px;
}

body {
  background-color: $teal;
  color: $dark-purple;
  font-family: "Montserrat", "Roboto", -apple-system, BlinkMacSystemFont, sans-serif;
  margin: 0;
  padding: 0;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

*:focus {
   outline: none;
}

.main-background {
  background-color: $light-purple;
  margin: 0;
  min-height: 415px;
  padding: 30px 2em;

  .main-container {
    margin: 0 auto;
  }
}

// iPhone X
@media only screen and (min-height: 812px) {

  .main-background {
    min-height: 550px;
  }
}

// iPad
@media only screen and (min-width: 768px) {

  .main-background {
    min-height: 700px;

    .main-container {
      max-width: 708px;
    }
  }
}

// iPad Pro
@media only screen and (min-height: 1366px) {

  .main-background {
    min-height: 1100px;
  }
}