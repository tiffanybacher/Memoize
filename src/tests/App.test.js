import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import App from '../components/App';

const getSavedCardSelection = jest.fn();
const getSavedCards = jest.fn();

const mockAllCards = [ 
  {
    "question": "What is an object in JavaScript?",
    "answer": "An object in JavaScript can be compared to an object in real life. It is a stand alone entity with it's own set of properties and methods.",
    "id": 1,
    "group": 1
  },
  {
    "question": "What is a method?",
    "answer": "A method is a function that describes a behavior of an object.",
    "id": 2,
    "group": 1
  }
];
const mockCardSelection = mockAllCards;
const mockCurrentCardData = {
    "question": "What is an object in JavaScript?",
    "answer": "An object in JavaScript can be compared to an object in real life. It is a stand alone entity with it's own set of properties and methods.",
    "id": 1,
    "group": 1
  }
const mockSavedCards = [mockCurrentCardData];
const mockQuestion = 'What is an object in JavaScript?';
const mockAnswer = 'An object in JavaScript can be compared to an object in real life. It is a stand alone entity with it\'s own set of properties and methods.';
const mockUserAnswer = 'An object is something that represents a thing in real life.';

const defaultState = {
  group: 1,
  allCards: '',
  cardSelection: null,
  currentCardData: '',
  answerIsShown: false,
  userAnswer: '',
  savedMsgIsShown: false,
  savedCards: null
}

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <App />
    );
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a default state', () => {
    expect(wrapper.state()).toEqual(defaultState);
  });

  it('should show answer when showAnswer is called', () => {
    wrapper.instance().showAnswer();

    expect(wrapper.state('answerIsShown')).toEqual(true);
  });

  it('should update userAnswer when getUserAnswer is called', () => {
    wrapper.instance().getUserAnswer(mockUserAnswer);

    expect(wrapper.state('userAnswer')).toEqual(mockUserAnswer);
  });

  it('should get saved card selection from local storage', () => {
    wrapper.instance().getSavedCardSelection();

    expect(localStorage.getItem).toHaveBeenCalled();
  });

  it('should start a fresh card selection using all the cards if no selection is found in local storage', () => {
    wrapper.setState({
      allCards: mockAllCards,
      cardSelection: null
    });

    wrapper.instance().startFreshCardSelection();

    expect(wrapper.state('cardSelection')).toEqual(mockAllCards);
  });

  it('should remove the current card of allCards and return a new selection to cardSelection if there is no current selection', () => {
    wrapper.setState({
      allCards: mockAllCards,
      cardSelection: null
    });

    wrapper.instance().updateCardSelection();

    expect(wrapper.state('cardSelection')).toEqual([{
      "question": "What is a method?",
      "answer": "A method is a function that describes a behavior of an object.",
      "id": 2,
      "group": 1
    }]);
  });

  it('should remove the current card of cardSelection and return a new selection to cardSelection', () => {
    wrapper.setState({
      allCards: mockAllCards,
      cardSelection: mockCardSelection
    });

    wrapper.instance().updateCardSelection();

    expect(wrapper.state('cardSelection')).toEqual([{
      "question": "What is a method?",
      "answer": "A method is a function that describes a behavior of an object.",
      "id": 2,
      "group": 1
    }]);
  });

  it('should save current card selection to storage if a current selection exists', () => {
    wrapper.instance().saveCurrentCardsToStorage();

    expect(localStorage.getItem).toHaveBeenCalled();

    expect(localStorage.setItem).toHaveBeenCalled();
  });

  it('should update currentCardData with the first item in cardSelection', () => {
    wrapper.setState({
      allCards: mockAllCards,
      cardSelection: mockCardSelection
    });

    wrapper.instance().updateCardData();

    expect(wrapper.state('currentCardData')).toEqual(mockCurrentCardData);
  });

  it('should update currentCardData with the first item in allCards if cardSelection is null', () => {
    wrapper.setState({
      allCards: mockAllCards,
      cardSelection: null
    });

    wrapper.instance().updateCardData();
    
    expect(wrapper.state('currentCardData')).toEqual(mockCurrentCardData);
  });

  it('should get save cards from local storage', () => {
    wrapper.instance().getSavedCards();

    expect(localStorage.getItem).toHaveBeenCalled();
  });

  it('should save card to local storage', () => {
    wrapper.instance().saveCardToStorage();

    expect(localStorage.setItem).toHaveBeenCalled();
  });

  it('should update savedCards when a card is saved', () => {
    wrapper.setState({
      currentCardData: mockCurrentCardData,
      savedCards: null
    });

    wrapper.instance().updateSavedCards();

    expect(wrapper.state('savedCards')).toEqual([mockCurrentCardData]);
  });

  it('should be able to take in multiple cards to savedCards', () => {
    wrapper.setState({
      currentCardData: mockCardSelection[1],
      savedCards: mockSavedCards
    });

    wrapper.instance().updateSavedCards();

    expect(wrapper.state('savedCards')).toEqual(mockCardSelection);
  });

  it('should hide answer', () => {
    wrapper.instance().hideAnswer();

    expect(wrapper.state('answerIsShown')).toEqual(false);
  });

  it('should show message letting user know we saved a card', () => {
    wrapper.instance().showSavedMsg();

    expect(wrapper.state('savedMsgIsShown')).toEqual(true);
  });

  it('should hide message letting user know we saved a card', () => {
    wrapper.instance().hideSavedMsg();

    expect(wrapper.state('savedMsgIsShown')).toEqual(false);
  });
});


