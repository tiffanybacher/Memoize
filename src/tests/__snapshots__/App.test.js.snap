// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <Nav />
  </header>
  <Cat />
  <section
    className="main-background"
  >
    <div
      className="main-container"
    >
      <Instructions
        answerIsShown={false}
        savedMsgIsShown={false}
      />
      <Flashcard
        answerIsShown={false}
      />
      <AnswerInput
        getUserAnswer={[Function]}
        showAnswer={[Function]}
      />
    </div>
  </section>
</div>
`;
